### PR TITLE
WT-6901 Fill content for the cursor subpage in the architecture guide

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
Can you please review this, @hbokhari?  There's a fair amount of content.

**Comment by hbokhari** [comment]
@ddanderson - I agree - there is lot of content to go through. I'll try to add comments as I read through the text.

File: src/docs/arch-cursor.dox [hunk 1]
```diff
@@ -3,13 +3,393 @@
 Cursors are used in WiredTiger to get and modify data.
 A caller of WiredTiger uses WT_SESSION::open_cursor to create
 a WT_CURSOR.  Methods on the WT_CURSOR can then be used to
-position, iterate, get, and set data.
+position, iterate, get, and set data.  In the typical case, a cursor
+will be used to access keys and values in a Btree.  Cursors are
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:7]: However, cursors can also be used to access indexed data, WiredTiger statistics, log files,
and metadata.

File: src/docs/arch-cursor.dox [hunk 2]
```diff
@@ -3,13 +3,393 @@
...
+also used to retrieve indexed data as well as WiredTiger statistics, log files,
+and metadata. Additionally, cursors are used for managing backups.
 
-Depending on the <code>uri</code> used when creating a cursor, the cursor will
-be internally implemented as one of the many cursor structures that include
-WT_CURSOR_BTREE, WT_CURSOR_BACKUP, WT_CURSOR_INDEX, WT_CURSOR_LOG,
-WT_CURSOR_METADATA, WT_CURSOR_STAT. Each of these structures starts
+The various kinds of cursors are created during the WT_SESSION::open_cursor call.
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:11]: during the -> using the

File: src/docs/arch-cursor.dox [hunk 3]
```diff
@@ -3,13 +3,393 @@
...
+Thus, any one of these "extended cursor structs" can be allocated and
+returned as a WT_CURSOR pointer.  Since the method pointers are filled with
+a specific implementations, for example \c __curtable_reset, a call to
+WT_CURSOR::reset will call the specific implementation function, and the first
+argument can be cast to be a pointer to the cursor struct used by the implementation.
+Thus, in our C code, we have something similar to a class hierarchy, having
+an abstract base class (WT_CURSOR) with virtual methods, and a number of
+implementation classes.
+
+The code for each cursor type's methods are generally grouped together,
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:27]: The code for each cursor type's methods is generally organized in a single file.

File: src/docs/arch-cursor.dox [hunk 4]
```diff
@@ -3,13 +3,393 @@
...
+for example, a backup cursor is implemented in \c src/cursor/cur_backup.c .
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:28]: . For example,

File: src/docs/arch-cursor.dox [hunk 5]
```diff
@@ -3,13 +3,393 @@
...
+Shared or utility cursor methods are in \c src/cursor/cur_std.c .
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:29]: Similarly, shared or utility

Comment (hbokhari) [reply src/docs/arch-cursor.dox:29]: are defined in \c

File: src/docs/arch-cursor.dox [hunk 6]
```diff
@@ -3,13 +3,393 @@
...
+
+Several cursor types have the concept of subordinate cursors, or child
+cursors.  For example, a table cursor is composed of subordinate file cursors,
+each representing a column group.  A metadata cursor has a subordinate
+cursor that is a file cursor on the metadata file.
+
+Every open cursor appears on a list in \c WT_SESSION->cursors .  This list is ordered
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:36]: Remove space after WT_SESSION->cursors

File: src/docs/arch-cursor.dox [hunk 7]
```diff
@@ -3,13 +3,393 @@
...
+in the way that cursors are closed when the session is closed.  Thus, if a subordinate
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:37]: in the way -> in a way

File: src/docs/arch-cursor.dox [hunk 8]
```diff
@@ -3,13 +3,393 @@
...
+cursor needs to be closed before its parent, it must be listed before the parent.
+
+@section arch_cursor_raw Data translation
+
+Cursors that expose Btree data, like file, table and index cursors, return a set of keys and values,
+translating encoded data to types that match the schema.  For example, if the \c value_format specified for
+a table is \c "iSi" (an integer, a null-terminated string, an integer), then data `{0, "abc", 5}` might
+be stored in the B-tree in a packed format as:
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:45]: For the sake of the consistency across document:
B-tree -> Btree

File: src/docs/arch-cursor.dox [hunk 9]
```diff
@@ -3,13 +3,393 @@
...
+
+A table cursor is a higher level concept built on file cursors.  A WiredTiger table allows data
+to be physically split into separate Btrees, this is done via the concept of column groups.  Column
+groups may be defined that contain a set of named columns (`columns` are synonymous with `fields`).
+Each column group's columns are the stored in a single btree and may be looked up by the table's key.
+See @ref schema_column_groups for API details.
+
+Tables may also have indices. These are implemented as Btrees mapping the index key(s) to the
+main key(s) of the table.  Indices may be added after the data is populated in the main table,
+this requires the index to be filled at the time it is added.  Index cursors are implemented using
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:77]: Use a conjunction rather than a comma.

File: src/docs/arch-cursor.dox [hunk 10]
```diff
@@ -3,13 +3,393 @@
...
+- \c "s" means skip a value, that is \c "x3".
+- \c "n" means get the next value, that is \c "x4".
+- \c ...
+
+Notice that this plan requires many switches of cursors and several \c "s" (skip) operations.
+Each skip involves enough decoding of the data item to determine its length so its data can
+be skipped over.
+
+With the default (complete) projection, getting values is fast.  Using the default plan, the needed
+columns are pulled out of each subordinate cursor one by one, being copied to the caller's arguments.
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:144]: being copied -> and get copied

File: src/docs/arch-cursor.dox [hunk 11]
```diff
@@ -3,13 +3,393 @@
...
+With projections, the simple algorithm following the plan works well if the columns in the projection
+are grouped by column group and requested in order.  Without that discipline, as in this example,
+the performance will not be optimal.
+
+The implementation of plan creation and execution resides the schema.
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:149]: resides the scheme -> resides in the schema.

File: src/docs/arch-cursor.dox [hunk 12]
```diff
@@ -3,13 +3,393 @@
...
+
+@section arch_cursor_dump Dump cursors
+
+Dump cursors are used in two rather different ways.  Regular dump cursors retrieve the raw keys and values
+and translate bytes either as raw characters or translating to hex.  This flavor of dump cursor is used
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:154]: translating to hex -> or as hex values

File: src/docs/arch-cursor.dox [hunk 13]
```diff
@@ -3,13 +3,393 @@
...
+by the `wt dump` or `wt dump -x` utility.  JSON dump cursors do more sophisticated translation, returning a string that is a JSON
+formatted record, with the name for each key and its data and the name for each value and its corresponding data.
+Data is translated to either integral, floating point, or string depending on the format of the column in the schema.
+This flavor of dump cursor is used by the `wt dump -j` command as part of creating a JSON dump of WiredTiger URI.
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:158]: of **_the_** WiredTiger URI

File: src/docs/arch-cursor.dox [hunk 14]
```diff
@@ -3,13 +3,393 @@
...
+
+There is a single \c WT_CURSOR_DUMP struct that is used to implement both flavors.
+The dump code checks for the \c WT_CURSTD_DUMP_JSON flag and as needed, calls into functions
+in `src/cursor/cur_json.c` .  The code in that file also implements several external functions that are used
+by the `wt` utility when loading JSON-dumped files.  In particular, the \c __wt_json_token function
+returns individual JSON tokens from an input string.
+
+The JSON code used by the dump cursor uses some storage that hangs off of `cursor->json_private` that
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:166]: that is types as -> which is typed as

File: src/docs/arch-cursor.dox [hunk 15]
```diff
@@ -3,13 +3,393 @@
...
+is typed as \c WT_CURSOR_JSON.  When a JSON flavored cursor is created, the list of key column names
+and value column names is populated in \c WT_CURSOR_JSON.  These names, obtained from the
+configuration string that created the table or file, are useful to have in advance,
+as they are used once per row to help fill out the JSON output.  The functions that get
+rows know to iterate these names and unpack the corresponding column data,
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:171]: know to iterate -> know how to iterate

File: src/docs/arch-cursor.dox [hunk 16]
```diff
@@ -3,13 +3,393 @@
...
+backup.  The backup cursor, when opened, ensures that it is the only backup cursor running
+in the system and returns an error if not.  This is managed using the
+\c WT_CONNECTION_IMPL->hot_backup_start variable, which can only be accessed when the
+connection schema lock is held.  A non-zero value means a hot backup is in progress.
+Closing the backup cursor sets it back to zero.
+
+Having an open backup influences actions
+elsewhere in WiredTiger, since part of the backup protocol involves the application copying
+whole data files.  Thus, having an open backup may cause the block manager and log file server
+thread to avoid truncating data and log files.  A truncate of a file being copied by the
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:190]: A truncate of file  -> truncation of a file

File: src/docs/arch-cursor.dox [hunk 17]
```diff
@@ -3,13 +3,393 @@
...
+application would be unexpected. Also, open checkpoints are not deleted during the
+course of a backup.
+
+When the backup cursor is initialized, the complete set of files needed to back up is generated
+and stored in the cursor.  This makes the backup's \c next function easy, just return the next
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:195]: , just return the next -> just returning the next file file in the list.

File: src/docs/arch-cursor.dox [hunk 18]
```diff
@@ -3,13 +3,393 @@
...
+an index's participation in one clause are stored in \c WT_CURSOR_JOIN_ENTRY objects.  \c WT_CURSOR_JOIN and
+\c WT_CURSOR_JOIN_ENTRY objects can be composed in a hierarchical manner, representing the shape of the tree representing
+the query.  These two structures are somewhat static, being created when the join is created.
+
+The join cursor contains a pointer to a \c WT_CURSOR_JOIN_ITER, which in some sense encodes the "position"
+of the cursor.  To get the next row that satisfies a join requires that multiple cursors be iterated.
+Generally, the "left-most" index cursor is iterated first.  Using the example above, the \c price
+index would be iterated, skipping over any entries that did not satisfy `(row.price > 100 and row.price <= 200)`.
+Then other cursors are checked to see that any other conditions are satisfied.
+The join may be configured to use Bloom filters, when that occurs,
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:225]: when that occurs -> and when that occurs

**Comment by hbokhari** [comment]
@ddanderson I have reviewed more than half of the document. Hopefully, I'll finish the review by tomorrow.

File: src/docs/arch-cursor.dox (hunk 2 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox:11]: I went with "by the"

File: src/docs/arch-cursor.dox (hunk 3 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox:27]: done.

File: src/docs/arch-cursor.dox (hunk 8 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox:45]: Fixed.  I've tried to use "Btree", that seems to be pretty common in the doc, though other variants like B-tree are used sometimes.

File: src/docs/arch-cursor.dox (hunk 15 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox:171]: I removed "know to", so it's just `The functions that get rows iterate these names...`. Seems clearer.

File: src/docs/arch-cursor.dox (hunk 9 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox]: I made these two sentences rather than use a conjunction or comma.

File: src/docs/arch-cursor.dox (hunk 13 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox]: Changed this to `JSON dump of a WiredTiger table or file`, I hope that's clearer.

**Comment by ddanderson** [comment]
Thanks @hbokhari, I've resolved your review comments.  I've replied and left comments unresolved when I've made adjustments to your suggestions.

File: src/docs/arch-cursor.dox [hunk 19]
```diff
@@ -3,13 +3,393 @@
...
+There is one \c WT_CURSOR_JOIN_ITER object corresponding to each level of nesting (queries can have
+arbitrary nesting of \c AND/OR).  Each \c WT_CURSOR_JOIN_ITER allows the question to be asked: does
+the current position of the main table cursor satisfy the join conditions for this part of the join tree?
+
+Part of the reason for this dynamic structure is that disjunctions may require some dynamic action.
+Consider
+```
+(row.price > 100 and row.in_stock < 10) .... or (row.aisle == 12 and row.on_sale == 1)
+```
+The last part of the query (being part of an OR clause) may be first executed after a number of items
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:239]: may be -> will probably be

File: src/docs/arch-cursor.dox [hunk 20]
```diff
@@ -3,13 +3,393 @@
...
+(*strong* reference) drops to zero, it means that no cursor is open on the dhandle, and the only
+references are from cached cursors. In this state, the dhandle may be marked dead by the dhandle sweep.
+When the dhandle is dead, the dhandle's memory will still persist, but each session will eventually
+notice, during its cursor cache sweep, and "fully close" the cursor, removing it from the cache list,
+releasing its weak reference before freeing the cursor.  Each session holding cached cursors must have
+some periodic activity that causes it to run its sweep, an occasional call to WT_SESSION::reset will suffice.
+After a dhandle has been dead for enough time, it is expected that all of its weak references will drop to zero,
+and the dhandle itself can be freed by the dhandle sweep.
+
+If there is a failure during the \c cache function, then we would want fully close the cursor.
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:317]: would want fully - would want to fully

File: src/docs/arch-cursor.dox [hunk 21]
```diff
@@ -3,13 +3,393 @@
...
+Rather than have special case code to handle this rare condition, we instead call \c reopen to
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:318]: rather than hav special -> rather than having

File: src/docs/arch-cursor.dox [hunk 22]
```diff
@@ -3,13 +3,393 @@
...
+In fact, if a WT_SESSION::drop call is called, we want to ensure that the data handle has been
+closed and the corresponding file is removed.  Some systems, like Windows, require that all open
+file handles be closed before a file can be successfully removed from the file system.  So there
+is motivation to periodically mark data handles that have no active references, and have sessions
+free cached cursors that have weak references to such marked data handles.  The former job
+occurs in the connection sweep code.  The latter job occurs in the session cursor cache sweep.
+
+The session cursor cache sweep currently happens in the WT_CURSOR::close call,
+and also on calls to WT_SESSION::reset.  On one hand, we don't want the overhead
+of a sweep to occur too frequently. It quite possible that both close and reset can be called a lot
```
Comment (hbokhari) [inline src/docs/arch-cursor.dox:355]: It quite possible -> it is quite possible

**Review by hbokhari — APPROVED** [review]
I have left few comments but overall I liked reading this document - well written and very detail oriented.

LGTM!

File: src/docs/arch-cursor.dox (hunk 19 above)
Comment (ddanderson) [reply src/docs/arch-cursor.dox:239]: Changed this from "may be" to "will be".

**Pull request merged**